numpyro/infer/mcmc.py

```diff
@@ -733,75 +742,127 @@ def _hashable(x):
+            to `False`.
+        :param init_params: Initial parameters to begin sampling. The type must be consistent
+            with the input type to `potential_fn`.
+        :param kwargs: Keyword arguments to be provided to the :meth:`numpyro.infer.mcmc.MCMCKernel.init`
+            method. These are typically the keyword arguments needed by the `model`.
+        """
+        if collect_warmup:
+            self._set_fori_collect_infos(0, self.num_warmup, self.num_warmup)
+        else:
+            self._set_fori_collect_infos(self.num_warmup, self.num_warmup, self.num_samples)
+        self.run(rng_key, *args, extra_fields=extra_fields, init_params=init_params, **kwargs)
+        self._warmup_state = self._last_state
 
-    def run(self, rng_key, *args, extra_fields=(), collect_warmup=False,
-            init_params=None, reuse_warmup_state=False, **kwargs):
+    def run(self, rng_key, *args, extra_fields=(), init_params=None, reuse_warmup_state=False, **kwargs):
         """
         Run the MCMC samplers and collect samples.
 
         :param random.PRNGKey rng_key: Random number generator key to be used for the sampling.
+            For multi-chains, a batch of `num_chains` keys can be supplied. If `rng_key`
+            does not have batch_size, it will be splitted in to a batch of `num_chains` keys.
         :param args: Arguments to be provided to the :meth:`numpyro.infer.mcmc.MCMCKernel.init` method.
             These are typically the arguments needed by the `model`.
         :param extra_fields: Extra fields (aside from `z`, `diverging`) from :data:`numpyro.infer.mcmc.HMCState`
             to collect during the MCMC run.
         :type extra_fields: tuple or list
-        :param bool collect_warmup: Whether to collect samples from the warmup phase. Defaults
-            to `False`.
         :param init_params: Initial parameters to begin sampling. The type must be consistent
             with the input type to `potential_fn`.
         :param bool reuse_warmup_state: If `True`, sampling would make use of the last state and
-            adaptation parameters from the previous warmup run.
+            adaptation parameters from the previous :meth:`warmup` run.
         :param kwargs: Keyword arguments to be provided to the :meth:`numpyro.infer.mcmc.MCMCKernel.init`
             method. These are typically the keyword arguments needed by the `model`.
         """
         self._args = args
         self._kwargs = kwargs
+        init_state = self._get_cached_init_state(rng_key, args, kwargs)
+        if self.num_chains > 1 and rng_key.ndim == 1:
+            rng_key = random.split(rng_key, self.num_chains)
+
+        if reuse_warmup_state:
+            self._set_fori_collect_infos(0, self.num_samples, self.num_samples)
+            if self._warmup_state is None:
+                raise ValueError('No `init_state` found; warmup results cannot be reused.')
```

Comment:
nit: change this to `_warmup_state` to reflect the actual attribute name. Also move up before `self._set_fori_..`.